script: improve spec compliance of the iframe load event steps

gterzian opened the pull request:
Prevent the running of the iframe load event steps for the initial about:blank document. Part of https://github.com/servo/servo/issues/31973

Testing: /html/browsers/browsing-the-web/navigating-across-documents/initial-empty-document/iframe-nosrc.html
Fixes: It is expected that the intermittent issue of 34819 will go away, but I will leave it open so we can close it when we realize it isn't being mentioned by PR's anymore. 

https://github.com/servo/servo/issues/34819

gterzian committed:
prevent running of iframe load event steps for initial about:blank docs

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fmt

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
always still terminate load blocker

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/html/htmliframeelement.rs:
```diff
@@ -530,6 +530,21 @@ impl HTMLIFrameElement {
+        // like any other document(and this is hard to refactor).
+        // this flag is necessary to prevent the load event steps from running.
+        if self.is_initial_blank_document() {
+            return;
```
Thanks, however at least one of the test that is timing out is still doing that even with such a change, but I'll give it a full run just to see if it makes a difference overall.

gterzian committed:
run load event steps for initial frame if it does not have an src

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
undo test expectation changes

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove prints

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add pending nav flag

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
condition load event steps for the initial blank document on the lack of pending nav and src

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
remove fail expectation for form submission of iframe-nosrc

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
undo test changes

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
clean-up and add docs

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
escape raw link

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

Taym95 commented on components/script/dom/html/htmliframeelement.rs:
```diff
@@ -84,6 +84,9 @@ pub(crate) struct HTMLIFrameElement {
     #[conditional_malloc_size_of]
     script_window_proxies: Rc<ScriptWindowProxies>,
+    /// Keeping track of whether the iframe will be navigated;
+    /// see the note in `iframe_load_event_steps`.
```
can we be more specific with this doc commnent, since this flag only means "a real navigation (has already been triggered while we're still on the initial about:blank" ?

Taym95 commented on components/script/dom/html/htmlformelement.rs:
```diff
@@ -1060,6 +1061,17 @@ impl HTMLFormElement {
 
+        // Note the pending form navigation if this is an iframe;
+        // necessary for deciding whether to run the iframe load event steps.
+        if let Some(window_proxy) = target.undiscarded_window_proxy() {
```
can you please confirm my understanding of this code, we mark the frame as having a pending navigation so that its about:blank load doesn't fire ?  Without this, we'dd dispatch load for about:blank and then again for the real navigation, which is not what browsers do and causes Servo issue??

gterzian replied:
Yes that's basically it. When a navigation has started, script expects the load event to fire on the result of the navigation, but since the blank doc could still be pending it's own async load event, this flag ensure those do not fire. If the load event had fired synchronously, this wouldn't be a problem because it would have fired already before script could navigate it.

Taym95 reviewed (approved):
LGTM, mostly nits and questions

gterzian committed:
improved docs

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
@taym95 Thanks for the review, I have addressed your comments by trying to clarify the docs. 

So this is really more of a patch to ensure the behavior from the WPT perspective improves, but internally it still doesn't follow the spec, although it does document the non-standard behavior and its relationship to the standard. 

But since further refactoring could break all kinds of other things, I prefer to take a small win now before attempting further changes. 

I've [noted some follow-ups](https://github.com/servo/servo/issues/31973#issuecomment-3467523140).

cc @jdm @mrobinson

gterzian commented:
> > I think what I really need to do is simply to run the load events steps synchronously for the initial blank document, like in the spec, as opposed to asynchronously via the constellation, because with the async approach there seems to be so many conditions to maintain.
> 
> See if there's anything you would like to reuse from [441266a](https://github.com/servo/servo/commit/441266a0007edd688aedc438a88e253444b57446) and [28c27ff](https://github.com/servo/servo/commit/28c27ff83527d7426199529227b9783afd4f95c2) which implemented that.

I've looked at the commits, and since they seem to run the load event steps in a microtask I think they are just as liable to run into the same issues as the current setup which happens via messaging with the constellation(the timing is likely to be different however). 

I think the solution is to literally try to follow the spec and run the steps synchronously when processing the iframe attributes for the first time(in the case of the blank document that stays on about:blank).

gterzian committed:
fmt

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>
